plugin/pkg/admission/ingressclass/admission_test.go

```diff
@@ -0,0 +1,196 @@
+	classWithNoDefault := &networkingv1.IngressClass{
+		TypeMeta: metav1.TypeMeta{
+			Kind: "IngressClass",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "nondefault2",
+		},
+	}
+	// Class with empty default annotation (=non-default)
+	classWithEmptyDefault := &networkingv1.IngressClass{
+		TypeMeta: metav1.TypeMeta{
+			Kind: "IngressClass",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "nondefault2",
+			Annotations: map[string]string{
+				networkingv1.AnnotationIsDefaultIngressClass: "",
+			},
+		},
+	}
+
+	testCases := []struct {
+		name            string
+		classes         []*networkingv1.IngressClass
+		classField      *string
+		classAnnotation *string
+		expectedClass   *string
+		expectedError   error
+	}{
+		{
+			name:            "no default, no modification of Ingress",
+			classes:         []*networkingv1.IngressClass{classWithFalseDefault, classWithNoDefault, classWithEmptyDefault},
+			classField:      nil,
+			classAnnotation: nil,
+			expectedClass:   nil,
+			expectedError:   nil,
+		},
+		{
+			name:            "one default, modify Ingress with class=nil",
+			classes:         []*networkingv1.IngressClass{defaultClass1, classWithFalseDefault, classWithNoDefault, classWithEmptyDefault},
+			classField:      nil,
+			classAnnotation: nil,
+			expectedClass:   utilpointer.StringPtr(defaultClass1.Name),
+			expectedError:   nil,
+		},
+		{
+			name:            "one default, no modification of Ingress with class field=''",
+			classes:         []*networkingv1.IngressClass{defaultClass1, classWithFalseDefault, classWithNoDefault, classWithEmptyDefault},
+			classField:      utilpointer.StringPtr(""),
+			classAnnotation: nil,
+			expectedClass:   utilpointer.StringPtr(""),
+			expectedError:   nil,
+		},
+		{
+			name:            "one default, no modification of Ingress with class field='foo'",
+			classes:         []*networkingv1.IngressClass{defaultClass1, classWithFalseDefault, classWithNoDefault, classWithEmptyDefault},
+			classField:      utilpointer.StringPtr("foo"),
+			classAnnotation: nil,
+			expectedClass:   utilpointer.StringPtr("foo"),
+			expectedError:   nil,
+		},
+		{
+			name:            "one default, no modification of Ingress with class annotation='foo'",
+			classes:         []*networkingv1.IngressClass{defaultClass1, classWithFalseDefault, classWithNoDefault, classWithEmptyDefault},
+			classField:      nil,
+			classAnnotation: utilpointer.StringPtr("foo"),
+			expectedClass:   nil,
+			expectedError:   nil,
+		},
+		{
+			name:            "two defaults, error with Ingress with class field=nil",
+			classes:         []*networkingv1.IngressClass{defaultClass1, defaultClass2, classWithFalseDefault, classWithNoDefault, classWithEmptyDefault},
+			classField:      nil,
+			classAnnotation: nil,
+			expectedClass:   nil,
+			expectedError:   errors.NewForbidden(networkingv1.Resource("ingresses"), "testing", errors.NewInternalError(fmt.Errorf("2 default IngressClasses were found, only 1 allowed"))),
```

Comment:
tests can become fragile if you test for the error message

Reply:
That's definitely true, in this case I thought it was worthwhile though. This is an admission controller, and it's important not just that it prevents invalid configuration, but that it includes a clear explanation for why the configuration is invalid. In this case, this is testing an error message that is specific to this admission controller, something that would only change if the admission controller itself changed.